api/desecapi/models/mfa.py

```diff
@@ -30,20 +32,35 @@ def _secret_default():
 
     secret = models.BinaryField(max_length=32, default=_secret_default.__func__)
 
+    @cached_property
+    def _totp(self):
+        # TODO switch to self.secret once https://github.com/pyauth/pyotp/pull/138 is released
+        return TOTP(self.base32_secret, digits=6)
+
     @property
     def base32_secret(self):
         return base64.b32encode(self.secret).rstrip(b"=").decode("ascii")
 
-    def verify(self, code):
-        # TODO switch to self.secret once https://github.com/pyauth/pyotp/pull/138 is released
-        totp = TOTP(self.base32_secret, digits=6)
+    @property
+    def uri(self):
+        return self._totp.provisioning_uri(
+            name=self.name,
+            issuer_name=f"desec.{settings.DESECSTACK_DOMAIN}",
+        )
 
+    def verify(self, code):
         now = datetime.now()
-        match_current = pyotp_utils.strings_equal(str(code), str(totp.at(now)))
-        match_prev = pyotp_utils.strings_equal(str(code), str(totp.at(now, -1)))
-        steps_ago = totp.timecode(now) - totp.timecode(self.last_used)
+        match_current = pyotp_utils.strings_equal(str(code), str(self._totp.at(now)))
+        match_prev = pyotp_utils.strings_equal(str(code), str(self._totp.at(now, -1)))
+        if self.last_used is None:
+            steps_ago = None
+        else:
+            steps_ago = self._totp.timecode(now) - self._totp.timecode(self.last_used)
         # Replay protection. Does not seem compatible with using old code followed by new code in the same timeslice.
-        if (steps_ago > 0 and match_current) or (steps_ago > 1 and match_prev):
+        # TODO save last used timecode
+        if (match_current and (steps_ago is None or steps_ago > 0)) or (
+            match_prev and (steps_ago is None or steps_ago > 1)
+        ):
```

Comment:
Why bother with the difference of `timecode(now)` and `timecode(last_used)`? The condition could read:

```python
(match_current and (last_used_step is None or current_step > last_used_step)) or
(match_prev and (last_used_step is None or prev_step > last_used_step))
```

A more explicit variant of the code that checks code and replay:

```python
matches = match_current or match_prev
if not matches:
    return False  # code does not match

matching_step = timecode(now) - match_prev  # lol maybe this is inappropriately ugly

if last_used_step and matching_step <= last_used_step:
    return False  # code was already used
```

No stakes, just writing this out to make sure your code does what I think it does. :)

Reply:
> Why bother with the difference of `timecode(now)` and `timecode(last_used)`?

Because then only one variable was needed, not two.

> A more explicit variant of the code that checks code and replay:

Almost. I don't think it addresses the case correctly when the code is the same consecutively (which is a 1e-6 chance, so not too low).

Anyway, I'm going to store the last used timecode in the database, and then all of this goes away.

Reply:
done, 681433788306909a5b6412b671dd87bdc6c23c28 and 2a1fb77be23fbc49d82df81e6a426e955de4dc07